### PR TITLE
Add checks to ensure existing PVC matches PVC to restore

### PR DESCRIPTION
**Pull request by ShyamsundarR**
Currently, the check for existing PVC, if found during PVC restore, checks to see if PVC exists and is restored by Ramen.

This is insufficient in the case of a StatefulSet where the PVC maybe left behind (similar to the PV) post a failover or a relocate and on restore of the PVC if the older PVC is found, it may not be restored by Ramen.

The checks for existing PVCs is hence expanded to ensure:
- PVC is not deleted
- PVC is bound
- PVC is bound to the same PV volume name we would have restored

The above ensures that we would have restored the same PVC and it would have been bound to the PV as desired.

Also, the check to ensure that if the PVC exists it is restored by Ramen is dropped, as that may fail in cases as mentioned above.

### CODE REVIEW

**Review by BenamarMk — APPROVED** [review]

**Review by raghavendra-talur — CHANGES_REQUESTED** [review]

File: controllers/vrg_volrep.go [hunk 1]
```diff
@@ -2024,9 +2024,24 @@ func (v *VRGInstance) validateExistingPVC(pvc *corev1.PersistentVolumeClaim) err
...
+		return fmt.Errorf("PVC exists and is not bound (phase: %s)", existingPVC.Status.Phase)
+	}
+
+	// If terminating return error
+	if !existingPVC.DeletionTimestamp.IsZero() {
+		return fmt.Errorf("existing bound PVC %s deletion timestamp non-zero", existingPVC.GetName())
+	}
+
+	// Check match to the PV we expect PVC to be bound to
+	if existingPVC.Spec.VolumeName != pvc.Spec.VolumeName {
```
Comment (ShyamsundarR) [inline controllers/vrg_volrep.go:2038]: Had changed this from == to != and forgot to return an error, fixing the same as well.

Comment (ShyamsundarR) [reply controllers/vrg_volrep.go:2039]: Addressed

File: controllers/vrg_volrep.go [hunk 2]
```diff
@@ -2024,9 +2024,24 @@ func (v *VRGInstance) validateExistingPVC(pvc *corev1.PersistentVolumeClaim) err
 		return fmt.Errorf("failed to get existing PVC (%w)", err)
 	}
 
-	if existingPVC.ObjectMeta.Annotations == nil ||
-		existingPVC.ObjectMeta.Annotations[RestoreAnnotation] != RestoredByRamen {
-		return fmt.Errorf("found PVC object not restored by Ramen for PVC %s", existingPVC.Name)
+	// If not Bound return error
+	if existingPVC.Status.Phase != corev1.ClaimBound {
+		return fmt.Errorf("PVC exists and is not bound (phase: %s)", existingPVC.Status.Phase)
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/vrg_volrep.go]: Addressed

**Comment by hatfieldbrian** [comment]
Applied this patch and it fixes #1009

File: controllers/vrg_volrep.go [hunk 3]
```diff
@@ -2018,20 +2018,32 @@ func (v *VRGInstance) validateExistingPV(pv *corev1.PersistentVolume) error {
...
 
-	if existingPVC.ObjectMeta.Annotations == nil ||
-		existingPVC.ObjectMeta.Annotations[RestoreAnnotation] != RestoredByRamen {
-		return fmt.Errorf("found PVC object not restored by Ramen for PVC %s", existingPVC.Name)
+	// If PVC not "Bound" return error
+	if existingPVC.Status.Phase != corev1.ClaimBound {
+		return fmt.Errorf("PVC %s exists and is not bound (phase: %s)", pvcNSName.String(), existingPVC.Status.Phase)
+	}
+
+	// If PVC is terminating return error
```
Comment (nirs) [inline controllers/vrg_volrep.go:2037]: The comment does not add anything to the code. Maybe we need a helper so we can say:

   if pvcs.isDeleted(existingPVC) {

File: controllers/vrg_volrep.go [hunk 4]
```diff
@@ -2018,20 +2018,32 @@ func (v *VRGInstance) validateExistingPV(pv *corev1.PersistentVolume) error {
...
+	err := v.reconciler.Get(v.ctx, pvcNSName, existingPVC)
 	if err != nil {
-		return fmt.Errorf("failed to get existing PVC (%w)", err)
+		return fmt.Errorf("failed to get existing PVC %s (%w)", pvcNSName.String(), err)
 	}
 
-	if existingPVC.ObjectMeta.Annotations == nil ||
-		existingPVC.ObjectMeta.Annotations[RestoreAnnotation] != RestoredByRamen {
-		return fmt.Errorf("found PVC object not restored by Ramen for PVC %s", existingPVC.Name)
+	// If PVC not "Bound" return error
```
Comment (nirs) [inline controllers/vrg_volrep.go:2032]: The comment does not add much. Maybe we need a helper function so we can say:

    if !pvcs.IsBound(existingPVC) {

File: controllers/vrg_volrep.go [hunk 5]
```diff
@@ -2018,20 +2018,32 @@ func (v *VRGInstance) validateExistingPV(pv *corev1.PersistentVolume) error {
...
+	if existingPVC.Status.Phase != corev1.ClaimBound {
+		return fmt.Errorf("PVC %s exists and is not bound (phase: %s)", pvcNSName.String(), existingPVC.Status.Phase)
+	}
+
+	// If PVC is terminating return error
+	if !existingPVC.DeletionTimestamp.IsZero() {
+		return fmt.Errorf("existing bound PVC %s is being deleted", pvcNSName.String())
+	}
+
+	// Check match to the PV we expect PVC to be bound to
```
Comment (nirs) [inline controllers/vrg_volrep.go:2042]: Same, the comment explain the code which is pretty clear so it can be removed.

Comment (ShyamsundarR) [reply controllers/vrg_volrep.go]: Removed the 3 comments...

File: controllers/vrg_volrep.go [hunk 6]
```diff
@@ -2018,20 +2018,32 @@ func (v *VRGInstance) validateExistingPV(pv *corev1.PersistentVolume) error {
 
 func (v *VRGInstance) validateExistingPVC(pvc *corev1.PersistentVolumeClaim) error {
 	existingPVC := &corev1.PersistentVolumeClaim{}
+	pvcNSName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
 
-	err := v.reconciler.Get(v.ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, existingPVC)
+	err := v.reconciler.Get(v.ctx, pvcNSName, existingPVC)
 	if err != nil {
-		return fmt.Errorf("failed to get existing PVC (%w)", err)
+		return fmt.Errorf("failed to get existing PVC %s (%w)", pvcNSName.String(), err)
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/vrg_volrep.go:2029]: The string for the type [NamespacedName](https://github.com/kubernetes/apimachinery/blob/v0.26.3/pkg/types/namespacedname.go#L27) return it in the string format "namespace/name" which is typical for a resource specification.

The fmt package contains [Stringer](https://pkg.go.dev/fmt#Stringer) that uses the method, but other packages/interfaces may not. Prefer to leave it explicit here.

[the remark this answers is not in the document]
Comment (nirs) [reply controllers/vrg_volrep.go:2029]: But we are dealing here with the fmt package - any type implementing the the Stringer interface works fine with `%s`, and this is very common pattern in many languages (.e.g. java toString(), python __str__()).

What is the benefit using explicit `.String()`?

[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/vrg_volrep.go:2029]: All structs do not have String method, hence want to leave this explicit. Also similar usage in the related [function](https://github.com/RamenDR/ramen/blob/d3e48b9855b16b6aae2f383d84411eb6ca53c533/controllers/vrg_volrep.go#L1980-L1981) above.

[the remark this answers is not in the document]
Comment (nirs) [reply controllers/vrg_volrep.go:2029]: Do you mean that some struct do not have String()?

So for structs that do not have a String() method we can use:

     fmt.Sprintf("%s/%s", s.Member1, s.Member2)

But when a struct has a String method that we can use:

     fmt.Sprintf("%s", s)

And we a struct is missing String method, we can send a PR to add it.

Keeping the current convention in the code around is good reason. We can remove
the unneeded String() in the entire module/package/project later.

**Pull request merged**